### ISSUE TITLE
RIS Gain on the Coverage Map, but not in sim_ber

### ISSUE THREAD

**Issue by ahmadfaisal9900** [NVlabs/sionna#988]
Hi, I'm trying to run an experiment where I'm trying to see the effects of binarizing the phases and making a end-to-end differentiable model using an RIS, for which I'm using Sionna v0.19.2.

I'm running into a weird problem, where a coverage map is showing me significant gains after optimization using Gradient Descent however upon training a NRX/comparing with Perfect-CSI, I do not observe any gains compared to the unoptimized RIS setup.

Scene: 
`scene = load_scene(file_path)
# Configure antenna array for all transmitters
scene.tx_array = PlanarArray(1,1,0.5,0.5,"iso", "V")
scene.rx_array = PlanarArray(1,1,0.5,0.5,"iso", "V")

# Create transmitter
tx = Transmitter(name="tx",
                 position=[-15,3,2])

# Add transmitter
scene.add(tx)

# Create a receiver
rx = Receiver(name="rx",
              position=[-2,6.5,2],)

# Add receiver
scene.add(rx)

width = 8 * scene.wavelength # Width [m] for 16*16 RIS elements
num_rows = num_cols = int(width/(0.5*scene.wavelength))
ris = RIS(name="ris",
          position=[-2,-2,2],
          orientation=[PI/2,0,0],
          num_rows=200,
          num_cols=200,
          num_modes=1,
          look_at=(tx.position + rx.position)/2
          )

scene.add(ris)
scene.frequency = 5e9`

Here's what the setup looks like: 

<img width="684" height="574" alt="Image" src="https://github.com/user-attachments/assets/8bdb553d-7bd8-4a4f-8d4f-078e3c4d8a3d" />

Training configuration: 

`# SNR range for evaluation and training
ebno_db_min = 0
ebno_db_max = 10

## OFDM configuration
subcarrier_spacing = 30e3 # Hz
fft_size = 128 # Number of subcarriers
num_ofdm_symbols = 14 # Number of OFDM symbols
dc_null = True # Null the DC subcarrier
num_guard_carriers = [5, 6] # Number of guard carriers on each side
pilot_pattern = "kronecker" # Pilot pattern
pilot_ofdm_symbol_indices = [2, 11] # Index of OFDM symbols carrying pilots
#pilot_ofdm_symbol_indices = [2] # Index of OFDM symbols carrying pilots
cyclic_prefix_length = 0 # Simulation in frequency domain.

## Modulation and coding configuration
num_bits_per_symbol = 2 # QPSK
coderate = 0.5 # Coderate for LDPC code

## Neural receiver configuration
num_conv_channels = 12 # Number of convolutional channels for the convolutional layers forming the neural receiver

## Training configuration
training_batch_size = 64 # Training batch size
model_weights_path = "neural_receiver_weights" # Location to save the neural receiver weights once training is done
stream_manager = StreamManagement(np.array([[1]]), # Receiver-transmitter association matrix
                                  1)               # One stream per transmitter
resource_grid = ResourceGrid(num_ofdm_symbols = num_ofdm_symbols,
                             fft_size = fft_size,
                             subcarrier_spacing = subcarrier_spacing,
                             num_tx = 1,
                             num_streams_per_tx = 1,
                             cyclic_prefix_length = cyclic_prefix_length,
                             dc_null = dc_null,
                             pilot_pattern = pilot_pattern,
                             pilot_ofdm_symbol_indices = pilot_ofdm_symbol_indices,
                             num_guard_carriers = num_guard_carriers)
# Codeword length. It is calculated from the total number of databits carried by the resource grid, and the number of bits transmitted per resource element
n = int(resource_grid.num_data_symbols*num_bits_per_symbol)
# Number of information bits per codeword
k = int(n*coderate)
BER = {}
paths = scene.compute_paths(max_depth=5, num_samples=1e6, reflection=True, ris=True)

a, tau = paths.cir()
# Compute frequencies of subcarriers and center around carrier frequency
frequencies = subcarrier_frequencies(fft_size, subcarrier_spacing)

# Compute the frequency response of the channel at frequencies.
h_freq = cir_to_ofdm_channel(frequencies,
                             a,
                             tau,
                             normalize=True)

binary_source = BinarySource()
mapper = Mapper("qam", num_bits_per_symbol)
rg_mapper = ResourceGridMapper(resource_grid)
batch_size = 64
ebno_db = tf.fill([batch_size], 0.0)
no = ebnodb2no(ebno_db, num_bits_per_symbol, coderate)
c = binary_source([batch_size, 1, 1, n])
x = mapper(c)
x_rg = rg_mapper(x)
channel = ApplyOFDMChannel(add_awgn=True)
`

Here's the E2ESystemModel:

`class E2ESystem(Model):

    def __init__(self, system, training=False):
        super().__init__()
        self._system = system
        self._training = training

        ######################################
        ## Transmitter
        self._binary_source = BinarySource()
        if not training:
            self._encoder = LDPC5GEncoder(k, n)
        self._mapper = Mapper("qam", num_bits_per_symbol)
        self._rg_mapper = ResourceGridMapper(resource_grid)
        ######################################
        ## Channel
        self._channel = ApplyOFDMChannel(add_awgn=True)
        ######################################
        ## Receiver
        if "baseline" in system:
            if system == 'baseline-perfect-csi':  # Perfect CSI
                self._removed_null_subc = RemoveNulledSubcarriers(resource_grid)
            elif system == 'baseline-ls-estimation':  # LS estimation
                self._ls_est = LSChannelEstimator(resource_grid, interpolation_type="nn")
            # Components required by both baselines
            self._lmmse_equ = LMMSEEqualizer(resource_grid, stream_manager)
            self._demapper = Demapper("app", "qam", num_bits_per_symbol)

        elif system == "neural-receiver":  # Neural receiver
            self._neural_receiver = NeuralReceiver()
            self._rg_demapper = ResourceGridDemapper(resource_grid, stream_manager)  # Used to extract data-carrying resource elements
        if not training:
            self._decoder = LDPC5GDecoder(self._encoder, hard_out=True)
    @tf.function
    def call(self, batch_size, ebno_db):

        # If `ebno_db` is a scalar, a tensor with shape [batch size] is created as it is what is expected by some layers
        if len(ebno_db.shape) == 0:
            ebno_db = tf.fill([batch_size], ebno_db)
        no = ebnodb2no(ebno_db, num_bits_per_symbol, coderate)
        if self._training:
            c = self._binary_source([batch_size, 1, 1, n])
        else:
            b = self._binary_source([batch_size, 1, 1, k])
            c = self._encoder(b)
        # Modulation
        x = self._mapper(c)
        x_rg = self._rg_mapper(x)
        no_ = expand_to_rank(no, tf.rank(x_rg))
        # channel
        y   = self._channel([x_rg,h_freq, no_])
        if "baseline" in self._system:
            if self._system == 'baseline-perfect-csi':
                h_hat = self._removed_null_subc(h_freq) # Extract non-null subcarriers
                batch_size = tf.shape(y)[0]  # Or pass batch_size explicitly if available
                # Tile h_hat across the batch dimension
                h_hat = tf.tile(h_hat, [batch_size, 1, 1, 1, 1, 14, 1])
                err_var = 0.0 # No channel estimation error when perfect CSI knowledge is assumed
            elif self._system == 'baseline-ls-estimation':
                h_hat, err_var = self._ls_est([y, no]) # LS channel estimation with nearest-neighbor

            x_hat, no_eff = self._lmmse_equ([y, h_hat, err_var, no]) # LMMSE equalization
            no_eff_= expand_to_rank(no_eff, tf.rank(x_hat))
            llr = self._demapper([x_hat, no_eff_]) # Demapping
        elif self._system == "neural-receiver":
            # The neural receiver computes LLRs from the frequency domain received symbols and N0
            y = tf.squeeze(y, axis=1)
            llr = self._neural_receiver([y, no])
            llr = insert_dims(llr, 2, 1) # Reshape the input to fit what the resource grid demapper is expected
            llr = self._rg_demapper(llr) # Extract data-carrying resource elements. The other LLrs are discarded
            llr = tf.reshape(llr, [batch_size, 1, 1, n]) # Reshape the LLRs to fit what the outer decoder is expected
        if self._training:
            # Compute and return BMD rate (in bit), which is known to be an achievable
            # information rate for BICM systems.
            # Training aims at maximizing the BMD rate
            bce = tf.nn.sigmoid_cross_entropy_with_logits(c, llr)
            bce = tf.reduce_mean(bce)
            rate = tf.constant(1.0, tf.float32) - bce/tf.math.log(2.)
            return rate
        else:
            # Outer decoding
            b_hat = self._decoder(llr)
            return b,b_hat # for BER/BLER computation
`
Baseline Comparision Results:

`num_training_iterations = 300 # Number of training iterations
training = True
if training:
    model = E2ESystem('neural-receiver', training=True)

    optimizer = tf.keras.optimizers.Adam()

    for i in range(num_training_iterations):
        # Sampling a batch of SNRs
        ebno_db = tf.random.uniform(shape=[], minval=ebno_db_min, maxval=ebno_db_max)
        # Forward pass
        with tf.GradientTape() as tape:
            rate = model(training_batch_size, ebno_db)
            loss = - rate
            #print(loss.numpy())

        # Computing and applying gradients
        weights = model.trainable_weights
        grads = tape.gradient(loss, weights)
        optimizer.apply_gradients(zip(grads, weights))
        if i % 100 == 0:
            print('Iteration {}/{}  Rate: {:.4f} bit'.format(i, num_training_iterations, rate.numpy()), end='\r')

def save_weights(model, model_weights_path):
    weights = model.get_weights()
    with open(model_weights_path, 'wb') as f:
        pickle.dump(weights, f)
save_weights(model, model_weights_path)
# Range of SNRs over which the systems are evaluated
ebno_dbs = np.arange(ebno_db_min, # Min SNR for evaluation
                     ebno_db_max, # Max SNR for evaluation
                     0.5) # Step
# function to load and set weights of a model
def load_weights(model, model_weights_path):
    model(1, tf.constant(10.0, tf.float32))
    with open(model_weights_path, 'rb') as f:
        weights = pickle.load(f)
    model.set_weights(weights)

model = E2ESystem('baseline-perfect-csi')
ber,_ = sim_ber(model, ebno_dbs, batch_size=64, num_target_block_errors=100, max_mc_iter=100)
BER['baseline-perfect-csi'] = ber.numpy()

model = E2ESystem('baseline-ls-estimation')
ber,_ = sim_ber(model, ebno_dbs, batch_size=64, num_target_block_errors=100, max_mc_iter=100)
BER['baseline-ls-estimation'] = ber.numpy()

model = E2ESystem('neural-receiver')

# Run one inference to build the layers and loading the weights
model(1, tf.constant(10.0, tf.float32))
with open(model_weights_path, 'rb') as f:
    weights = pickle.load(f)
model.set_weights(weights)

# Evaluations
ber,_ = sim_ber(model, ebno_dbs, batch_size=64, num_target_block_errors=100, max_mc_iter=100)
BER['neural-receiver'] = ber.numpy()`

Here's my code for setting up the scene, and optimization via RIS:
`

# ──────────────────────────────────────────────────────────────
# 0.  Imports & common helpers
# ──────────────────────────────────────────────────────────────
import tensorflow as tf, numpy as np, matplotlib.pyplot as plt
from tqdm import tqdm
from sionna.rt.solver_paths import SolverPaths

PI = np.pi
tf.keras.backend.set_floatx("float32")

def to_db(x):
    return 10*tf.math.log(x + 1e-12) / tf.math.log(10.0)

# ──────────────────────────────────────────────────────────────
# 1.  RIS solver with in-solver STE binarisation (0 or π)
# ──────────────────────────────────────────────────────────────
class BinarizedSolverPaths(SolverPaths):
    """Identical to your CustomSolverPaths, except phases are forced
       to {0 , π} with a straight-through estimator."""
    def _ris_transition_matrices(self, ris_paths, ris_paths_tmp):
        # ---- (a) coefficient calculation – your original lines ----
        coef = tf.cast(4*PI, self._rdtype)
        coef /= tf.reduce_prod(ris_paths_tmp.distances, axis=0)
        coef *= tf.cast(tf.sqrt(tx.power_dbm * 0.001 * 120 * PI * 2 * 66),
                        self._rdtype)
        coef = tf.complex(coef, tf.cast(0, self._rdtype))

        # ---- (b) fetch phase vector --------------------------------
        phase_vec = tf.reshape(
            tf.cast(self._scene.ris["ris"].phase_profile.values,
                    self._rdtype), [-1])                   # (N_elems,)

        phase_rad = phase_vec

        # ---- (c) 1-bit straight-through projection -----------------
        hard_phase = tf.where(phase_rad >= 0.0,
                              tf.constant(np.pi, self._rdtype),
                              tf.constant(0.0 , self._rdtype))
        ste_phase  = phase_rad + tf.stop_gradient(hard_phase - phase_rad)

        # ---- (d) apply phase just like before ----------------------
        sqrt_sigma = tf.complex(tf.sqrt(tf.cast(0.001, self._rdtype)), 0.)
        coef = coef * sqrt_sigma * tf.exp(
            tf.complex(tf.zeros_like(ste_phase), ste_phase))

        # ---- (e) mask & 2×2 transition matrix ----------------------
        coef = tf.where(ris_paths.mask, coef, tf.cast(0.0, coef.dtype))
        coef = coef[..., tf.newaxis, tf.newaxis]
        return tf.cast(coef, self._dtype) * tf.eye(
            2, batch_shape=[1,1,1], dtype=self._dtype)

# ──────────────────────────────────────────────────────────────
# 2.  Attach solver to scene & make phase trainable
# ──────────────────────────────────────────────────────────────
scene._solver_paths = BinarizedSolverPaths(scene)

ris.phase_profile.values = tf.Variable(
    tf.zeros_like(ris.phase_profile.values), trainable=True)

# ──────────────────────────────────────────────────────────────
# 3.  Optimiser & single-step training function
# ──────────────────────────────────────────────────────────────
optimizer = tf.keras.optimizers.Adam(1e-2)

def train_step():
    with tf.GradientTape() as tape:
        # forward pass (binarisation happens inside solver)
        paths      = scene.compute_paths(max_depth=5,
                                         num_samples=int(5e6),
                                         reflection=True, ris=True)
        a, _       = paths.cir()                 # [1, n_rx, n_tx]
        power_lin  = tf.reduce_mean(tf.reduce_sum(tf.abs(a)**2, axis=-1))
        loss       = -power_lin                 # maximise power

    grads = tape.gradient(loss, [ris.phase_profile.values])
    optimizer.apply_gradients(zip(grads, [ris.phase_profile.values]))

    # optional: keep phases in [−π, π] for numeric stability
    ris.phase_profile.values.assign(
        tf.clip_by_value(ris.phase_profile.values, -PI, PI))

    return to_db(power_lin)

# ──────────────────────────────────────────────────────────────
# 4.  Training loop & simple plot
# ──────────────────────────────────────────────────────────────
num_iterations = 200
gain_hist = []

for i in tqdm(range(num_iterations)):
    gain_db = train_step().numpy()
    gain_hist.append(gain_db)
    if i % 10 == 0:
        print(f"Iter {i:3d} | Avg path gain = {gain_db:6.2f} dB")

plt.figure(figsize=(10,5))
plt.plot(gain_hist, 'r-o', ms=3)
plt.title("Average Path Gain over Iterations (1-bit RIS)")
plt.xlabel("Iteration"); plt.ylabel("Gain [dB]")
plt.grid(True); plt.tight_layout(); plt.show()

num_training_iterations = 300 # Number of training iterations
training = True
if training:
    model = E2ESystem('neural-receiver', training=True)

    optimizer = tf.keras.optimizers.Adam()

    for i in tqdm(range(num_training_iterations)):
        # Sampling a batch of SNRs
        ebno_db = tf.random.uniform(shape=[], minval=ebno_db_min, maxval=ebno_db_max)
        # Forward pass
        with tf.GradientTape() as tape:
            rate = model(training_batch_size, ebno_db)
            loss = - rate
            #print(loss.numpy())

        # Computing and applying gradients
        weights = model.trainable_weights
        grads = tape.gradient(loss, weights)
        optimizer.apply_gradients(zip(grads, weights))
        if i % 100 == 0:
            print('Iteration {}/{}  Rate: {:.4f} bit'.format(i, num_training_iterations, rate.numpy()), end='\r')

save_weights(model, model_weights_path)
model = E2ESystem('baseline-perfect-csi')
ber,_ = sim_ber(model, ebno_dbs, batch_size=64, num_target_block_errors=100, max_mc_iter=100)
BER['baseline-perfect-csi-PLOpRIS'] = ber.numpy()

model = E2ESystem('baseline-ls-estimation')
ber,_ = sim_ber(model, ebno_dbs, batch_size=64, num_target_block_errors=100, max_mc_iter=100)
BER['baseline-ls-estimation-PLOpRIS'] = ber.numpy()

model = E2ESystem('neural-receiver')

# Run one inference to build the layers and loading the weights
model(1, tf.constant(10.0, tf.float32))
with open(model_weights_path, 'rb') as f:
    weights = pickle.load(f)
model.set_weights(weights)

# Evaluations
ber,_ = sim_ber(model, ebno_dbs, batch_size=64, num_target_block_errors=100, max_mc_iter=100)
BER['neural-receiver-PLOpRIS'] = ber.numpy()
`

Here's the gains from this:

<img width="728" height="623" alt="Image" src="https://github.com/user-attachments/assets/fde8a4f6-b689-4481-a00b-dcef26143bc5" />

However, when I'm comparing the results of the sim_ber, I get absolutely no difference. 

Why is this a problem?

<img width="1218" height="688" alt="Image" src="https://github.com/user-attachments/assets/cdb9e6d1-11ea-4da2-beda-013727e9931d" />

I'm new to Sionna so this is probably something very fundamental that I'm missing, any help would be appreciated.

**Comment by ahmadfaisal9900** [comment]
Hi @jhoydis, thank you for pointing this out. 

I posted the snippet with the normalization. Without it, I get gains that aren't TOO great

<img width="718" height="635" alt="Image" src="https://github.com/user-attachments/assets/1c4c797c-0789-4641-abcb-3c6f4fc467d8" />

However, I get absolutely horrendous performance for sim_ber. Not sure if that's normal, as I get a really high BER even for 10db of E_b/N_0. I'll try to increase the transmit power and then see, perhaps that might be the issue.

**Comment by ahmadfaisal9900** [comment]
Nope, still pretty horrendous - even for a range of 15 to 30.

<img width="924" height="546" alt="Image" src="https://github.com/user-attachments/assets/cf33eaa7-cacd-4149-9346-63d83bc4a3a1" />

**Comment by jhoydis** [comment]
Hi @ahmadfaisal9900,

The reason for this is fairly simple. When you do not normalize the channel, you need to account for the pathloss and use a meaningful transmit power. In your simulations, the transmit power is equal to one and the noise power chosen to achieve a certain EbNo value under the assumption that the channel has unit energy. This is not the case anymore when you do not normalize the channel...
